ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs

```diff
@@ -442,53 +450,82 @@ validateHeader cfg ledgerView hdr st = do
   Serialisation
 -------------------------------------------------------------------------------}
 
-encodeAnnTip :: (HeaderHash blk -> Encoding)
-             -> (TipInfo    blk -> Encoding)
-             -> (AnnTip     blk -> Encoding)
-encodeAnnTip encodeHash encodeInfo AnnTip{..} = mconcat [
+defaultEncodeAnnTip :: TipInfo blk ~ HeaderHash blk
+                    => (HeaderHash blk -> Encoding)
+                    -> (AnnTip     blk -> Encoding)
+defaultEncodeAnnTip encodeHash AnnTip{..} = mconcat [
       encodeListLen 4
     , encode     annTipSlotNo
-    , encodeHash annTipHash
+    , encodeHash annTipInfo
     , encode     annTipBlockNo
-    , encodeInfo annTipInfo
+      -- TODO: Useless field. Remove when OK to break binary compatibility.
```

Comment:
How will we think of this when the opportunity is there to break binary compatibility?

Do we actually expect this to happen?

Alternatively, we do the simple naive thing in the encoder and make the decoder backwards compatible by branching on the size (4 -> old, 3 -> new).

Reply:
Once we move the hard fork combinator we break binary compatibility, I think that's the natural point to introduce this. When we create the tickets for the various TODOs in #2034 and create one to switch the main code base over to `HardForkBlock [Byron, Shelley]`, maybe let's change it then?

Reply:
https://github.com/input-output-hk/ouroboros-network/issues/2062